File: Grind75/47-TimeBasedKeyValueStore/python/TimeBasedKeyValueStore.py
#!/bin/python3

from __future__ import annotations
import sys
sys.path.append('../../Utility/python')
from typing import List, Dict, Tuple

class TimeMap:
    def __init__(self):
      self._tm: Dict[ str, List[Tuple[str,int]] ] = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
      if key not in self._tm:
        self._tm[key] = []
      self._tm[key].append(( value, timestamp ))

    def get(self, key: str, timestamp: int) -> str:
      if key not in self._tm:
        return ""
      values = self._tm[key]
      if timestamp < values[0][1]:
        return ""
      elif timestamp > values[-1][1]:
        return values[-1][0]
      left, right = 0, len(values)-1
      while left <= right:
        center = (left + right) // 2
        ts = values[center][1]
        if timestamp < ts:
          right = center - 1
        elif timestamp > ts:
          left = center + 1
        else:
          return values[center][0]
      return values[right][0]


if __name__ == '__main__':
  def quot(s: str): return "'" + s + "'"
  tm = TimeMap()
  tm.set("key", "LOW", 1)
  tm.set("key", "HIGH", 3)
  for i in range(5):
    print(i, quot(tm.get("key", i)))

File: Grind75/47-TimeBasedKeyValueStore/cpp/TimeBasedKeyValueStore.cpp
#include "TimeBasedKeyValueStore.h"
#include <string>
#include <utility>

using namespace std;

void TimeMap::set(string key, string value, int timestamp) {
  if (tm.find(key) == tm.end())
    tm[key] = vector<pair<string, int>>();
  tm[key].push_back({value, timestamp});
}

string TimeMap::get(string key, int timestamp) {
  if (tm.find(key) == tm.end())
    return "";
  vector<pair<string, int>> &timestamps = tm[key];
  if (timestamp < timestamps.front().second)
    return "";
  else if (timestamp > timestamps.back().second)
    return timestamps.back().first;
  int left = 0, right = timestamps.size() - 1;
  while (left <= right) {
    int center = (left + right) / 2;
    if (timestamp < timestamps[center].second)
      right = center - 1;
    else if (timestamp > timestamps[center].second)
      left = center + 1;
    else
      return timestamps[center].first;
  }
  return timestamps[right].first;
}

void TimeMap::set2(string key, string value, int timestamp) {
  if (timeMap.find(key) == timeMap.end())
    timeMap[key] = map<int, string>();
  timeMap[key][timestamp] = value;
}

string TimeMap::get2(string key, int timestamp) {
  if (timeMap.find(key) == timeMap.end())
    return "";
  map<int, string>::iterator found = timeMap[key].upper_bound(timestamp);
  if (found == timeMap[key].begin() && timestamp < found->first)
    return "";
  return (--found)->second;
}

